src/library_html5.js

```diff
@@ -672,9 +684,16 @@ var LibraryJSEvents = {
 
       var dpiScale = (strategy.canvasResolutionScaleMode == {{{ cDefine('EMSCRIPTEN_FULLSCREEN_CANVAS_SCALE_HIDEF') }}}) ? window.devicePixelRatio : 1;
       if (strategy.canvasResolutionScaleMode != {{{ cDefine('EMSCRIPTEN_FULLSCREEN_CANVAS_SCALE_NONE') }}}) {
-        target.width = cssWidth * dpiScale;
-        target.height = cssHeight * dpiScale;
-        if (target.GLctxObject) target.GLctxObject.GLctx.viewport(0, 0, target.width, target.height);
+        var newWidth = (cssWidth * dpiScale)|0;
+        var newHeight = (cssHeight * dpiScale)|0;
+
+        if (!target.controlTransferredOffscreen) {
```

Comment:
this pattern (if not control transferred, apply, othrewise do the call) appears 3 times. can the logic be moved into the call, that is, `emscripten_set_canvas_element_size_js` would check if the target has `controlTransferredOffscreen` and handle that?

Reply:
Really good point. Changed this, and while doing that, I realize that I am not actually ready to expose this function as a "public" function to JavaScript side, because that is something that will need a closer thought as part of the big elephant issue we have with the `JSEvents` object being a non-DCEable bloat (#2999, #3298). I think we need to remove that whole object and make all functions individually DCE well (even when Closure is not used), and that will require changes to how users expose the functions to JS code, if they do not call them from C++.

So for now changed the canvas set and get functions to be underscored internal functions - the above refactor would publicize them in a proper manner, while doing the same for other functions.